src/draw_target.rs

```diff
@@ -514,7 +515,11 @@ impl DrawState {
                 // subtract with overflow later.
                 usize::max(terminal_len, 1)
             };
-            if real_len + diff > term_height {
+            // Don't consider orphan lines when comparing to terminal height.
+            debug_assert!(idx <= real_len);
+            if self.orphan_lines_count <= idx
+                && real_len - self.orphan_lines_count + diff > term_height
```

Comment:
how does `idx` play into this? It just counts entries, not lines.

Reply:
Please forgive me, I don't know the precise meanings of these terms ("entries" and "lines").

But what I seem to observe is this. Following a call to `println` with a string containing `n` newlines, `self.orphan_lines_count` will be (at least) `n + 1`, and `self.lines` will contain (at least) `n + 1` elements.

The purpose of checking `idx` is see whether all of those orphan lines have been processed.

Reply:
Thanks for clarifying.

> we're trying to keep `real_len < term_height`

Is that really what we want, though? Not `real_len - self.orphan_lines_count < term_height`?

To be more precise, isn't it that we want to keep the vertical space of the displayed progress bars less than `term_height`?

> So I think your issue should also be fixable by comparing against `real_len` instead of `idx`

Here's why that suggestion gives me pause. Suppose we break with `idx < self.orphan_lines_count`. Then some number of orphan lines have not been drawn. In other words, the user called `println`, but some portion of the output has not been shown. Isn't that undesirable?

Wouldn't we want to show those lines (possibly allowing them to scroll off the top of the screen), and then draw just as many progress bars will fit?

Reply:
ah... right.

In that case, isn't even `idx` causing `println` lines to disappear?
If I understand you correctly, we'll always want to render orphan lines, and ignore them in the height math, acting as if they don't contribute to the terminal space that gets used up.

So... do we actually want `(real_len + diff).saturating_sub(self.orphan_lines_count) > term_height` without any other checks?

Reply:
@chris-laplante maybe you want to weigh in here as well?

Reply:
> If I understand you correctly, we'll always want to render orphan lines, and ignore them in the height math, acting as if they don't contribute to the terminal space that gets used up.

That's how I think of it, yes.

But I don't see how these follow.
> In that case, isn't even `idx` causing `println` lines to disappear?

> So... do we actually want `(real_len + diff).saturating_sub(self.orphan_lines_count) > term_height` without any other checks?

If we were try to attach "units" to these variables, `self.orphan_lines_count` and `idx` and would count "unformatted lines", but `real_len` would count "formatted (i.e,. narrower than the terminal) lines", correct?

So wouldn't `(real_len + diff).saturating_sub(self.orphan_lines_count)` be a kind of "type error," because it tries to add a formatted line count (`real_len + diff`) to an unformatted line count (`self.orphan_lines_count`)?

Reply:
> @chris-laplante maybe you want to weigh in here as well?

I'll try to take a look later this week, sorry, a little short on time for now :(

Reply:
@oli-obk Did you have any further thoughts on this?

Reply:
I assumed the `orphan_lines_count` was the real printed lines count. If that is not true, then your reasoning is correct and we need a way to fix that.

Reply:
To help increase confidence, I updated the `orphan_lines_message_above_progress_bar` test to generate orphan lines that wrap around the terminal.

From what I can tell, the fix seems to handle those lines correctly.